From ff31f516b3f44fdcf3cbdf68c18e0f36082e723d Mon Sep 17 00:00:00 2001
From: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
Date: Wed, 19 Apr 2023 12:43:48 +0000
Subject: [PATCH] MSE: limit the number of samples stored in SourceBuffer

Applications may append a lot of samples without triggereing
eviction algorithm (> 500K). Which results in cleanup(on player
destruction) taking too long time.

Here is an example of video where we observed such behavior:
https://www.youtube.com/tv#/watch?v=6ATfcsAFFNE&list=RD-WLgxoa4BeU

This change allows at least one pass of eviction algorithm if number
of samples is greater than 10K. Which should be enough to allow
storing ~2-3 minutes of audio or video if size limit doesn't kick in.

%% original patch: 1074.patch
---
 Source/WebCore/Modules/mediasource/SourceBuffer.cpp | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/Source/WebCore/Modules/mediasource/SourceBuffer.cpp b/Source/WebCore/Modules/mediasource/SourceBuffer.cpp
index 7b1dd5bcf012..0c8504b0de04 100644
--- a/Source/WebCore/Modules/mediasource/SourceBuffer.cpp
+++ b/Source/WebCore/Modules/mediasource/SourceBuffer.cpp
@@ -962,10 +962,18 @@ void SourceBuffer::evictCodedFrames(size_t newDataSize)
     if (isRemoved())
         return;
 
+    const auto hasTooManySamples = [&] {
+        size_t currentSize = 0;
+        for (const auto& trackBuffer : m_trackBufferMap.values())
+            currentSize += trackBuffer.samples.size();
+        const size_t kSampleMapSizeLimit = 10000;
+        return currentSize > kSampleMapSizeLimit;
+    };
+    
     // This algorithm is run to free up space in this source buffer when new data is appended.
     // 1. Let new data equal the data that is about to be appended to this SourceBuffer.
     // 2. If the buffer full flag equals false, then abort these steps.
-    if (!m_bufferFull)
+    if (!m_bufferFull && !hasTooManySamples())
         return;
 
     size_t maximumBufferSize = this->maximumBufferSize();
@@ -1006,7 +1014,8 @@ void SourceBuffer::evictCodedFrames(size_t newDataSize)
         removeCodedFrames(rangeStart, std::min(rangeEnd, maximumRangeEnd));
         if (extraMemoryCost() + newDataSize < maximumBufferSize) {
             m_bufferFull = false;
-            break;
+            if (!hasTooManySamples())
+                break;
         }
 
         rangeStart += thirtySeconds;
-- 
2.17.1

